File: fons/norma/hal/codegen/ts/toml.ts
/**
 * toml.ts - TOML Encoding/Decoding Implementation
 *
 * Native TypeScript implementation of the HAL TOML interface.
 * Uses Bun's built-in TOML.parse and smol-toml for stringify.
 *
 * Note: TOML root must be a table (object), not array or primitive.
 *
 * Verb meanings:
 *   - pange (compose): serialize table to TOML string
 *   - solve (untangle): parse TOML string to table
 *   - tempta (try): attempt to parse, return null on error
 */

import { stringify as smolStringify } from 'smol-toml';

// Bun has built-in TOML.parse
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const BunTOML = (globalThis as any).Bun?.TOML;

function parse(toml: string): unknown {
    if (BunTOML) {
        return BunTOML.parse(toml);
    }
    throw new Error('TOML parsing requires Bun runtime');
}

function stringify(value: unknown): string {
    if (typeof value !== 'object' || value === null || Array.isArray(value)) {
        throw new Error('TOML root must be a table (object)');
    }
    return smolStringify(value as Record<string, unknown>);
}

export const toml = {
    // =========================================================================
    // SERIALIZATION
    // =========================================================================

    /** Serialize table to TOML string */
    pange(valor: unknown): string {
        return stringify(valor);
    },

    // =========================================================================
    // PARSING
    // =========================================================================

    /** Parse TOML string to table (throws on error) */
    solve(tomlStr: string): unknown {
        return parse(tomlStr);
    },

    /** Attempt to parse TOML string (returns null on error) */
    tempta(tomlStr: string): unknown | null {
        try {
            return parse(tomlStr);
        }
        catch {
            return null;
        }
    },

    // =========================================================================
    // TYPE CHECKING
    // =========================================================================

    estNihil(valor: unknown): boolean {
        return valor === null || valor === undefined;
    },

    estBivalens(valor: unknown): boolean {
        return typeof valor === 'boolean';
    },

    estTextus(valor: unknown): boolean {
        return typeof valor === 'string';
    },

    estInteger(valor: unknown): boolean {
        return typeof valor === 'number' && Number.isInteger(valor);
    },

    estFractus(valor: unknown): boolean {
        return typeof valor === 'number' && !Number.isInteger(valor);
    },

    estTempus(valor: unknown): boolean {
        return valor instanceof Date;
    },

    estLista(valor: unknown): boolean {
        return Array.isArray(valor);
    },

    estTabula(valor: unknown): boolean {
        return typeof valor === 'object' && valor !== null && !Array.isArray(valor) && !(valor instanceof Date);
    },

    // =========================================================================
    // VALUE EXTRACTION
    // =========================================================================

    utTextus(valor: unknown, defVal: string): string {
        return typeof valor === 'string' ? valor : defVal;
    },

    utNumerus(valor: unknown, defVal: number): number {
        return typeof valor === 'number' ? valor : defVal;
    },

    utBivalens(valor: unknown, defVal: boolean): boolean {
        return typeof valor === 'boolean' ? valor : defVal;
    },

    // =========================================================================
    // VALUE ACCESS
    // =========================================================================

    /** Get value by key (returns null if missing) */
    cape(valor: unknown, clavis: string): unknown {
        if (typeof valor === 'object' && valor !== null && !Array.isArray(valor)) {
            return (valor as Record<string, unknown>)[clavis] ?? null;
        }
        return null;
    },

    /** Pluck value by array index (returns null if out of bounds) */
    carpe(valor: unknown, index: number): unknown {
        if (Array.isArray(valor) && index >= 0 && index < valor.length) {
            return valor[index];
        }
        return null;
    },

    /** Find value by dotted path (returns null if not found) */
    inveni(valor: unknown, via: string): unknown {
        const parts = via.split('.');
        let current: unknown = valor;

        for (const part of parts) {
            if (current === null || current === undefined) {
                return null;
            }
            if (typeof current === 'object' && !Array.isArray(current) && !(current instanceof Date)) {
                current = (current as Record<string, unknown>)[part];
            }
            else if (Array.isArray(current)) {
                const idx = parseInt(part, 10);
                if (isNaN(idx) || idx < 0 || idx >= current.length) {
                    return null;
                }
                current = current[idx];
            }
            else {
                return null;
            }
        }

        return current ?? null;
    },
};

File: runtimes/norma-go/toml/toml.go
// Package toml provides TOML encoding/decoding for the Faber HAL.
//
// Etymology: "TOML" - Tom's Obvious Minimal Language (universal, kept as-is).
package toml

import (
	"reflect"
	"strings"
	"time"

	"github.com/pelletier/go-toml/v2"
)

// =============================================================================
// SERIALIZATION
// =============================================================================

// Pange serializes table to TOML string.
func Pange(valor any) (string, error) {
	data, err := toml.Marshal(valor)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// =============================================================================
// PARSING
// =============================================================================

// Solve parses TOML string to table.
func Solve(tomlStr string) (any, error) {
	var result map[string]any
	err := toml.Unmarshal([]byte(tomlStr), &result)
	if err != nil {
		return nil, err
	}
	return result, nil
}

// Tempta attempts to parse TOML string, returns nil on error.
func Tempta(tomlStr string) any {
	result, err := Solve(tomlStr)
	if err != nil {
		return nil
	}
	return result
}

// =============================================================================
// TYPE CHECKING
// =============================================================================

// EstNihil checks if value is nil.
func EstNihil(valor any) bool {
	return valor == nil
}

// EstBivalens checks if value is a boolean.
func EstBivalens(valor any) bool {
	_, ok := valor.(bool)
	return ok
}

// EstTextus checks if value is a string.
func EstTextus(valor any) bool {
	_, ok := valor.(string)
	return ok
}

// EstInteger checks if value is an integer.
func EstInteger(valor any) bool {
	switch valor.(type) {
	case int, int64, int32, int16, int8, uint, uint64, uint32, uint16, uint8:
		return true
	default:
		return false
	}
}

// EstFractus checks if value is a float.
func EstFractus(valor any) bool {
	switch valor.(type) {
	case float64, float32:
		return true
	default:
		return false
	}
}

// EstTempus checks if value is a time/datetime.
func EstTempus(valor any) bool {
	_, ok := valor.(time.Time)
	return ok
}

// EstLista checks if value is an array/slice.
func EstLista(valor any) bool {
	if valor == nil {
		return false
	}
	v := reflect.ValueOf(valor)
	return v.Kind() == reflect.Slice || v.Kind() == reflect.Array
}

// EstTabula checks if value is a table/map.
func EstTabula(valor any) bool {
	if valor == nil {
		return false
	}
	v := reflect.ValueOf(valor)
	return v.Kind() == reflect.Map
}

// =============================================================================
// VALUE EXTRACTION
// =============================================================================

// UtTextus extracts string value or returns default.
func UtTextus(valor any, defVal string) string {
	if s, ok := valor.(string); ok {
		return s
	}
	return defVal
}

// UtNumerus extracts numeric value or returns default.
func UtNumerus(valor any, defVal int) int {
	switch v := valor.(type) {
	case float64:
		return int(v)
	case float32:
		return int(v)
	case int:
		return v
	case int64:
		return int(v)
	case int32:
		return int(v)
	default:
		return defVal
	}
}

// UtBivalens extracts boolean value or returns default.
func UtBivalens(valor any, defVal bool) bool {
	if b, ok := valor.(bool); ok {
		return b
	}
	return defVal
}

// =============================================================================
// VALUE ACCESS
// =============================================================================

// Cape gets value by key from table (returns nil if missing).
func Cape(valor any, clavis string) any {
	if m, ok := valor.(map[string]any); ok {
		return m[clavis]
	}
	return nil
}

// Carpe plucks value by index from array (returns nil if out of bounds).
func Carpe(valor any, index int) any {
	if arr, ok := valor.([]any); ok {
		if index >= 0 && index < len(arr) {
			return arr[index]
		}
	}
	return nil
}

// Inveni finds value by dotted path (returns nil if not found).
func Inveni(valor any, via string) any {
	parts := strings.Split(via, ".")
	current := valor

	for _, part := range parts {
		if current == nil {
			return nil
		}
		if m, ok := current.(map[string]any); ok {
			current = m[part]
		} else {
			return nil
		}
	}

	return current
}
